docs: add related work (#2333)
apache/arrow-adbc: docs/source/format/related_work.rst

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at
..
..   http://www.apache.org/licenses/LICENSE-2.0
..
.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

============
Related Work
============

In the initial proposal, a survey of existing solutions and systems was
included, which is partially reproduced below for context, though note the
descriptions are only kept up-to-date on a best-effort basis.

> Preexisting database client APIs
> =============================

Author: ```suggestion
Preexisting database client APIs
================================
```

:external:doc:`Arrow Flight SQL <format/FlightSql>`
  A standard building on top of Arrow Flight, defining how to use
  Flight to talk to databases, retrieve metadata, execute queries, and
  so on. Provides a single client in C++ and Java language that talks
  to any database servers implementing the protocol. Models its API
  surface (though not API design) after JDBC and ODBC.

`DBI for R <https://www.r-dbi.org/>`_
  An R package/ecosystem of packages for database access. Provides a
  single interface with "backends" for specific databases.  While
  row-oriented, `integration with Arrow is under consideration`_,
  including a sketch of effectively the same idea as ADBC.

`JDBC <https://jcp.org/en/jsr/detail?id=221>`_
  A Java library for database access, providing row-based
  APIs. Provides a single interface with drivers for specific
  databases.

`ODBC <https://github.com/microsoft/ODBC-Specification>`_
  A language-agnostic standard from the ISO/IEC for database access,
  associated with Microsoft. Feature-wise, it is similar to JDBC (and
  indeed JDBC can wrap ODBC drivers), but it offers columnar data

>   support through fetching buffers of column values. (See above for
>   caveats.) Provides a single C interface with drivers for specific
>   databases.

Reviewer: ```suggestion
  support through fetching buffers of column values (with some
  caveats). Provides a single C interface with drivers for specific
  databases.
```

`PEP 249 <https://www.python.org/dev/peps/pep-0249/>`_ (DBAPI 2.0)
  A Python standard for database access providing row-based APIs. Not
  a singular package, but rather a set of interfaces that packages
  implement.

> Existing libraries

Reviewer: ```suggestion
Preexisting libraries
```

> Preexisting libraries
> ==================

Author: ```suggestion
Preexisting libraries
=====================
```
